File: backend/site_fit/contracts.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


SITE_FIT_CONTRACT_VERSION = "site_fit_contract_v1"


class SiteFitBaseRequest(BaseModel):
    plan: Optional[dict] = None
    structure: Optional[dict] = None
    site_constraints: dict
    design_locks: dict = Field(default_factory=dict)
    jurisdiction: Optional[str] = None
    ruleset_version: str = SITE_FIT_CONTRACT_VERSION


class SiteFitAnalyzeRequest(SiteFitBaseRequest):
    pass


class SiteFitApplyRequest(SiteFitBaseRequest):
    candidate_id: str


class SiteFitIsolationResponse(BaseModel):
    pipeline: str = "site_fit"
    separate_contracts: bool = True
    touched_existing_parse_generate_pipeline: bool = False


class SiteFitPlanSummaryResponse(BaseModel):
    source_kind: str
    canonical_unit: Optional[str] = None
    room_count: int = 0
    wall_count: int = 0
    opening_count: int = 0
    footprint_bbox: Optional[dict] = None


class SiteFitRegistrationTransformResponse(BaseModel):
    scale: float = 1.0
    rotation_degrees: float = 0.0
    translate_x: float = 0.0
    translate_y: float = 0.0


class SiteFitRegistrationSummaryResponse(BaseModel):
    status: str
    canonical_unit: Optional[str] = None
    scale_locked: bool = True
    transform: SiteFitRegistrationTransformResponse = Field(default_factory=SiteFitRegistrationTransformResponse)
    registered_plan_bbox: Optional[dict] = None
    warnings: list[str] = Field(default_factory=list)


class SiteFitComplianceSummaryResponse(BaseModel):
    status: str
    checked_rule_ids: list[str] = Field(default_factory=list)
    violations: list[dict] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class SiteFitCandidateResponse(BaseModel):
    candidate_id: str
    strategy: str
    summary: str
    fit_status: str
    change_count: int = 0
    score: float = 0.0
    changes: list[dict] = Field(default_factory=list)


class SiteFitAnalysisResponse(BaseModel):
    analysis_id: str
    contract_version: str = SITE_FIT_CONTRACT_VERSION
    status: str
    isolation: SiteFitIsolationResponse
    plan_summary: SiteFitPlanSummaryResponse
    registration_summary: SiteFitRegistrationSummaryResponse
    site_summary: dict = Field(default_factory=dict)
    compliance_summary: SiteFitComplianceSummaryResponse
    warnings: list[str] = Field(default_factory=list)


class SiteFitProposeResponse(BaseModel):
    analysis_id: str
    contract_version: str = SITE_FIT_CONTRACT_VERSION
    status: str
    isolation: SiteFitIsolationResponse
    plan_summary: SiteFitPlanSummaryResponse
    registration_summary: SiteFitRegistrationSummaryResponse
    site_summary: dict = Field(default_factory=dict)
    compliance_summary: SiteFitComplianceSummaryResponse
    candidates: list[SiteFitCandidateResponse] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class SiteFitApplyResponse(BaseModel):
    analysis_id: str
    contract_version: str = SITE_FIT_CONTRACT_VERSION
    candidate_id: str
    apply_status: str
    isolation: SiteFitIsolationResponse
    registration_summary: SiteFitRegistrationSummaryResponse
    compliance_summary: SiteFitComplianceSummaryResponse
    applied_plan: dict = Field(default_factory=dict)
    change_set: list[dict] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class SiteFitValidateResponse(BaseModel):
    analysis_id: str
    contract_version: str = SITE_FIT_CONTRACT_VERSION
    status: str
    isolation: SiteFitIsolationResponse
    plan_summary: SiteFitPlanSummaryResponse
    registration_summary: SiteFitRegistrationSummaryResponse
    site_summary: dict = Field(default_factory=dict)
    compliance_summary: SiteFitComplianceSummaryResponse
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/features/siteFit/contracts.ts
import type {
  CadReviewArtifactData,
  CadWorkspaceBBox,
  CadWorkspaceExtractResult,
  CadWorkspaceRoom,
} from '../cad/contracts'

export interface SiteFitCandidateChange {
  boundary_id: string
  side: string
  delta_x: number
  delta_y: number
  owner_room_ids?: string[]
  opening_ids?: string[]
  requires_rehost?: boolean
}

export interface SiteFitComplianceSummary {
  status: string
  checked_rule_ids: string[]
  violations: Array<{
    rule_id?: string
    message?: string
    reason?: string
  }>
  warnings: string[]
  boundary_diagnostics: Array<{
    boundary_id?: string
    status?: string
    reason?: string | null
    side?: string
    axis?: string
    overflow_delta?: number
    owner_room_ids?: string[]
    opening_ids?: string[]
    requires_rehost?: boolean
  }>
  room_diagnostics: Array<{
    room_id?: string
    boundary_id?: string
    status?: string
    reason?: string | null
    projected_width?: number
    projected_height?: number
    projected_area?: number
  }>
  mutation_hints: Array<{
    boundary_id?: string
    side?: string
    axis?: string
    delta_x?: number
    delta_y?: number
    owner_room_ids?: string[]
    opening_ids?: string[]
    requires_rehost?: boolean
  }>
}

export interface SiteFitRegistrationSummary {
  status: string
  canonical_unit?: string | null
  scale_locked?: boolean
  transform?: {
    scale?: number
    rotation_degrees?: number
    translate_x?: number
    translate_y?: number
  }
  registered_plan_bbox?: CadWorkspaceBBox | null
  warnings?: string[]
}

export interface SiteFitPlanSummary {
  source_kind: string
  canonical_unit?: string | null
  room_count: number
  wall_count: number
  opening_count: number
  footprint_bbox?: CadWorkspaceBBox | null
  movable_boundary_count?: number
  protected_boundary_count?: number
  locked_boundary_count?: number
  rehostable_opening_count?: number
}

export interface SiteFitProposalCandidate {
  candidate_id: string
  strategy: string
  summary: string
  fit_status: string
  change_count: number
  score?: number
  changes?: SiteFitCandidateChange[]
}

export interface SiteFitFootprintSummary {
  current?: CadWorkspaceBBox | null
  projected?: CadWorkspaceBBox | null
  buildable?: CadWorkspaceBBox | null
  widthDelta?: number | null
  heightDelta?: number | null
}

export interface SiteFitRoomMeasureSummary {
  name: string
  width: number
  height: number
  area: number
}

export interface SiteFitBridgeProposalResult {
  pipeline: 'site_fit_bridge_mvp_v1'
  scope: 'seminole-2000-only'
  plan_id: string
  plan_name: string
  cad_analysis: CadWorkspaceExtractResult
  proposal_review?: CadWorkspaceExtractResult
  site_constraints: Record<string, unknown>
  proposal: {
    analysis_id?: string
    status: string
    plan_summary?: SiteFitPlanSummary
    registration_summary?: SiteFitRegistrationSummary
    site_summary?: {
      buildable_bbox?: CadWorkspaceBBox | null
      has_buildable_envelope?: boolean
      locked_room_count?: number
      site_unit?: string
    }
    compliance_summary?: SiteFitComplianceSummary
    candidates?: Array<SiteFitProposalCandidate | null> | null
    warnings: string[]
  }
  warnings: string[]
}

export interface SiteFitBridgeApplyResult {
  pipeline: 'site_fit_bridge_mvp_v1'
  scope: 'seminole-2000-only'
  plan_id: string
  plan_name: string
  apply_id: string
  export_url: string
  applied_review: CadWorkspaceExtractResult
  apply: {
    candidate_id: string
    apply_status: string
    registration_summary?: SiteFitRegistrationSummary
    compliance_summary: { status: string }
    change_set?: SiteFitCandidateChange[]
    warnings: string[]
  }
  warnings: string[]
}

export interface SiteFitProposalArtifactData {
  planId: string
  planName: string
  candidateId: string | null
  cadAnalysisId: string | null
  siteConstraints: Record<string, unknown>
  summary: string
  fitStatus: string
  candidateStrategy?: string | null
  changeCount?: number
  preview?: CadReviewArtifactData
  footprint?: SiteFitFootprintSummary
  violationMessages?: string[]
  blockerMessages?: string[]
  mutationHintCount?: number
  changedRoomIds?: string[]
  warnings: string[]
}

export interface SiteFitApplyArtifactData {
  planId: string
  planName: string
  candidateId: string
  applyId: string
  applyStatus: string
  complianceStatus: string
  href?: string
  exportUrl?: string
  preview?: CadReviewArtifactData
  changeCount?: number
  changedRoomIds?: string[]
  beforeFootprint?: CadWorkspaceBBox | null
  afterFootprint?: CadWorkspaceBBox | null
  rooms?: SiteFitRoomMeasureSummary[]
  warnings: string[]
}
